subprojects/core/src/main/java/org/gradle/execution/TaskSelector.java

```diff
@@ -37,6 +38,7 @@
     private final GradleInternal gradle;
     private final ProjectConfigurer configurer;
     private final TaskPathResolver taskPathResolver = new TaskPathResolver();
+    private final BuildStateRegistry buildStateRegistry;
```

Comment:
I think we should avoid pushing more included build knowledge into this class and instead separate things a bit.

This might not be fully fleshed out, but here's a general idea:
https://github.com/gradle/gradle/commit/f3f961e4ff2f2015cdf5834169fac3fc7de8115b

In this change, the DefaultTaskSelector only knows how to select tasks from a single build.  A CompositeAwareTaskSelector knows about all of the available included builds and creates a selector for the build. This pulls the services up to the BuildTree level, which seems the right level to coordinate this vs each build reaching out to the build state registry. WDYT?

Reply:
The idea sounds goood, I just need to review your change before cherry-picking it to this PR. Working on it rn.

Reply:
I was more or less happy about my original design hiding behind `TaskPathResolver`, but the new organization of the services seems convincing. I'll cherry-pick your changes and check if I can do some cleanup.